04-rag-demo/README.md

```diff
@@ -0,0 +1,135 @@
+                    method: 'POST',
+                    headers: {
+                        'Authorization': `Bearer <your-openai-api-key>`,
+                        'Content-Type': 'application/json'
+                    },
+                    body: JSON.stringify({
+                        input: t,
+                        model: "text-embedding-3-small"
+                    })
+                });
+                if (!response.ok) {
+                    const error = await response.json();
+                    throw new Error(`OpenAI API error: ${error.error?.message || 'Unknown error'}`);
+                }
+                const data = await response.json();
+                return data.data[0].embedding;
+            } catch (error) {
+            }
+        }
+        return null;
+    }
+$$ WITH (
+    async = true,
+    batch = false
+);
+```
+
+Now, create the data pipeline to ingest the documents into the database. In RisingWave, a data pipeline is simply written as a materialized view defined in SQL.
+
+```sql
+CREATE TABLE documents (
+    file_name VARCHAR,
+    content TEXT,
+    PRIMARY KEY (file_name)
+);
+
+CREATE MATERIALIZED VIEW document_embeddings AS
+WITH t AS (
+    SELECT
+        *, text_embedding(content) AS embedding
+    FROM documents
+)
+SELECT
+    file_name,
+    content,
+    embedding
+FROM t
+WHERE embedding IS NOT NULL;
+```
+
+## Load data
+
+For this demo, we use the documents from the RisingWave docs. 
+
+```bash
+git clone https://github.com/risingwavelabs/risingwave-docs.git
+cd risingwave-docs
+
+# Load all the documents (`.mdx` files) into the `documents` table
+find . -name "*.mdx" | while read -r file; do
+    content=$(cat "$file" | sed "s/'/''/g")
+    sql_statement="INSERT INTO documents (file_name, content) VALUES ('$file', '$content') ON CONFLICT (file_name) DO UPDATE SET content = EXCLUDED.content;"
+    echo "$sql_statement" | psql -h 127.0.0.1 -p 4566 -d dev -U root
+    echo "Processed: $file"
+done
+```
+
+## Query
+
+To compare the similarity between the question and the documents, we need to introduce the `cosine_similarity` UDF.
+
+```sql
+create function cosine_similarity(v1 real[], v2 real[]) returns real language rust as $$
+    fn cosine_similarity(a: &[f32], b: &[f32]) -> f32 {
+        let dot_product: f32 = a.iter().zip(b).map(|(a, b)| a * b).sum();
+        let norm_a: f32 = a.iter().map(|a| a * a).sum();
+        let norm_b: f32 = b.iter().map(|b| b * b).sum();
+        dot_product / (norm_a * norm_b).sqrt()
+    }
+$$;
+```
+
+Now, we can use the `document_embeddings` materialized view to answer questions.
+
+The following SQL uses the `text_embedding` UDF to embed the question, and then finds the top 10 most similar documents from the `document_embeddings` materialized view.
+
+```sql
+WITH question_embedding AS (
+    SELECT text_embedding('write a Python UDF') as embedding
+    LIMIT 1 -- Hack: ensure the function is called once only
```

Comment:
What a hack...

Reply:
Can you please take a look someday? You'd better just run the query with & without `limit 1` locally. Without `limit 1`, our optimizer tend to pull up the UDF `Project` above the `Join`, which results a huge amplification.